libc/watcom/asm/segments.asm

```diff
@@ -25,61 +25,12 @@
 ;*
 ;*  ========================================================================
 ;*
-;* Description:  Watcom C segment ordering and null pointer section offsets
+;* Description:  Watcom C segment ordering for init/fini LIBC
 ;*
 ;*****************************************************************************
 
 
-        name    cstart
-        assume  nothing
-
- DGROUP group _NULL,_AFTERNULL,CONST,STRINGS,_DATA,DATA,XIB,XI,XIE,YIB,YI,YIE,_BSS
-
-; this guarantees that no function pointer will equal NULL
-; (WLINK will keep segment 'BEGTEXT' in front)
-; This segment must be at least 2 bytes in size to avoid confusing the
-; signal function.
-; need a symbol defined here to prevent the dead code elimination from
-; eliminating the segment.
-; (the int 3h is useful for quickly revealing jumps to NULL code pointers)
-
-BEGTEXT  segment word public 'CODE'
-        assume  cs:BEGTEXT
-        int     3
-        nop
-__begtext label byte
-        assume  cs:nothing
```

Comment:
I would prefer to keep the existing BEGTEXT segment as the int3/nop guarantees function pointers won't be NULL, as well as (int)function_ptr > 2 so that signals work (SIG_DFL, SIG_IGN are 0 and 1).

Likewise, the _NULL segment below is required so that zero-valued pointers don't arbitrarily scribble into the application data segment. This happens a lot on 8086 because there is no memory management.

Reply:
I understand, for OW version of libc it is possible to use it because wlink support these BEGTEXT, NULL and AFTERNULL segments and ensure they are on offset 0. But this guards is active only for near code/data.

Reply:
Please keep BEGTEXT, NULL, AFTERNULL etc as-is. If you think a better way of ensuring BEGTEXT is included than using `__begtext` (and/or `_cstart_`) that is OK. Deletion of segment definition rejected.

It is also OK for only small and medium models to work with near NULL pointer. I am not too worried about creating special handling for far-data programs, unless you have good idea for it.